rasa/utils/tensorflow/rasa_layers.py

```diff
@@ -0,0 +1,757 @@
+    ) -> tf.Tensor:
+        """Combine multiple 3-D dense/sparse feature tensors into one.
+
+        Arguments:
+            inputs: Tuple containing:
+                sequence_features: List of 3-D dense or sparse tensors with token-level
+                    features.
+                sentence_features: List of 3-D dense or sparse tensors with sentence-level
+                    features.
+                mask_sequence: a 3-D tensor mask that has 1s at real and 0s at padded
+                    positions corresponding to tokens in `sequence_features`.
+                mask_sequence_sentence: a 3-D tensor mask similar to `mask_sequence` but
+                    having each sequence of 1s longer by 1 to account for sequence 
+                    lengths of sequence- and sentence-level features being combined.
+            training: Python boolean indicating whether the layer should behave in
+                training mode (applying dropout to sparse tensors if applicable) or in
+                inference mode (not applying dropout).
+
+        Returns:
+            Single dense 3-D tensor containing the combined sequence- and sentence-
+            level, sparse & dense features.
+        """
+        sequence_features = inputs[0]
+        sentence_features = inputs[1]
+        mask_sequence = inputs[2]
+        mask_sequence_sentence = inputs[3]
+
+        # different feature types are present, make them dense & combine them
+        if self.have_all_feature_types:
+            sequence = self._tf_layers[f"sparse_dense.{SEQUENCE}"](
+                (sequence_features,), training=training
+            )
+
+            # apply mask which has 1s at positions of real tokens and 0s at all padded
+            # token positions. This is needed because the sparse+dense combining layer
+            # might've turned some fake (padded) features (i.e. 0s) into non-zero numbers
+            # and we want those to become zeros again.
+            # This step isn't needed for sentence-level features because those are never
+            # padded -- the effective sequence length in their case is always 1.
+            if mask_sequence is not None:
+                sequence = sequence * mask_sequence
+
+            sentence = self._tf_layers[f"sparse_dense.{SENTENCE}"](
+                (sentence_features,), training=training
+            )
+
+            sequence_sentence = self._tf_layers["concat_seq_sent"](
+                (sequence, sentence, mask_sequence_sentence)
+            )
+
+            return sequence_sentence
+
+        # only one feature type is present - make it dense but skip combining
+        elif f"sparse_dense.{SEQUENCE}" in self._tf_layers:
+            sequence = self._tf_layers[f"sparse_dense.{SEQUENCE}"](
+                (sequence_features,), training=training
+            )
+
+            return sequence
+        else:
+            sentence = self._tf_layers[f"sparse_dense.{SENTENCE}"](
```

Comment:
why `sparse_dense.{SEQUENCE}` might not exist but `sparse_dense.{SENTENCE}` always exists?

Reply:
It doesn't always exist. This:
```python
elif f"sparse_dense.{SEQUENCE}" in self._tf_layers:
```
is just a proxy for checking that only sequence-level feature signatures are present.

Reply:
isn't your concat layer capable of automatically figuring out whether to combine or not? Why don't you use it?

Reply:
True, true and true. Changing it now.